src/mlpack/tests/ann_layer_test.cpp

```diff
@@ -3810,4 +3810,166 @@ BOOST_AUTO_TEST_CASE(TransposedConvolutionalLayerOptionalParameterTest)
+
+  CheckMatrices(result, deterministicOutput, 1e-3);
+
+  // Check correctness by updating the running mean and variance again.
+  module1.Deterministic() = false;
+
+  // Clean up.
+  output.clear();
+  input.clear();
+
+  // Input consists of a single output of 2 feature maps with size 3 x 2.
+  input = arma::mat(6, 2);
+  input << 12 << 443 << arma::endr
+      << 134 << 45 << arma::endr
+      << 11 << 13 << arma::endr
+      << 14 << 55 << arma::endr
+      << 110 << 4 << arma::endr
+      << 1 << 45 << arma::endr;
+
+  result << -0.629337 << 2.14791 << arma::endr
+      << 0.156797 << -0.416694 << arma::endr
+      << -0.63578 << -0.622893 << arma::endr
+      << -0.637481 << 0.4440386 << arma::endr
+      << 1.894857 << -0.901267 << arma::endr
+      << -0.980402 << 0.180253 << arma::endr;
+
+  module1.Forward(input, output);
+  CheckMatrices(result, output, 1e-3);
+
+  runningMean(0) = 16.1792;
+  runningMean(1) = 6.30667;
+  runningVar(0) = 4278.545;
+  runningVar(1) = 201.78576;
+
+  CheckMatrices(runningMean, module1.TrainingMean(), 1e-3);
+  CheckMatrices(runningVar, module1.TrainingVariance(), 1e-3);
+
+  // Check backward function.
+  module1.Backward(input, output, delta);
+
+  deterministicOutput.clear();
+  module1.Deterministic() = true;
+  module1.Forward(input, deterministicOutput);
+
+  result.clear();
+  result << -0.06389128 << 6.525255983 << arma::endr
+      << 1.80125016 << 0.440614188 << arma::endr
+      << -0.07917933155 << -0.0486032421 << arma::endr
+      << 0.5415883265 << 3.427869269 << arma::endr
+      << 7.299709599 << -0.16238263 << arma::endr
+      << -0.3735739237 << 2.723898308 << arma::endr;
+
+  CheckMatrices(result, deterministicOutput, 1e-3);
+}
+
+/**
+ * Transposed Convolution layer numerical gradient test.
+ */
+BOOST_AUTO_TEST_CASE(GradientBatchNormWithMiniBatchesTest)
+{
+  // Add function gradient instantiation.
+  struct GradientFunction
+  {
+    GradientFunction()
+    {
+      input = arma::randn(16, 256);
+      arma::mat target;
+      target.ones(1, 256);
+
+      model = new FFN<NegativeLogLikelihood<>, NguyenWidrowInitialization>();
+      model->Predictors() = input;
+      model->Responses() = target;
+      model->Add<IdentityLayer<>>();
+      model->Add<Convolution<>>(1, 10, 3, 3, 1, 1, 0, 0, 4, 4);
+      //model->Add<BatchNorm<>>(10);
+      model->Add<Linear<>>(10 * 2 * 2, 2);
+      model->Add<LogSoftMax<>>();
+    }
+
+    ~GradientFunction()
+    {
+      delete model;
+    }
+
+    double Gradient(arma::mat& gradient) const
+    {
+      double error = model->Evaluate(model->Parameters(), 0, 256, false);
+      model->Gradient(model->Parameters(), 0, gradient, 256);
+      return error;
+    }
+
+    arma::mat& Parameters() { return model->Parameters(); }
+
+    FFN<NegativeLogLikelihood<>, NguyenWidrowInitialization>* model;
+    arma::mat input, target;
+  } function;
+
+  BOOST_REQUIRE_LE(CheckGradient(function), 0.4);
```

Comment:
Hey @saksham189, @KimSangYeon-DGU, The gradient threshold is already very high without the batch norm layer.  If I uncomment the batch norm layer, it will still pass the test. The minimum threshold required with just the convolutional layer is around 0.3. Could you please suggest how I should proceed?

Reply:
I think with reduced number of channels I got an error of 0.05. (So to keep it bit above that, I have set threshold as 0.08).

Reply:
I tried various combinations of the network, checked the implementation and I got a minimum error of 0.04. I have made all the other changes. Kindly let me know what you think.